src/topology/homotopy/homotopy_group.lean

```diff
@@ -154,42 +314,140 @@ begin
       prop' := λ _ _ ⟨i,_⟩, i.elim0 }⟩]
 end
 
-/-- The 1-dimensional generalized loops based at `x` are in 1-1 correspondence with
-  paths from `x` to itself. -/
-@[simps] def gen_loop_one_equiv_path_self : gen_loop 1 x ≃ path x x :=
-{ to_fun := λ p, path.mk ⟨λ t, p (λ _, t), by {continuity, exact p.1.2}⟩
-    (p.boundary (λ _, 0) ⟨0, or.inl rfl⟩)
-    (p.boundary (λ _, 1) ⟨1, or.inr rfl⟩),
+/-- The 1-dimensional generalized loops based at `x` are in 1-1 correspondence with loops at `x`. -/
+@[simps] def gen_loop_one_equiv_loop : Ω^(fin 1) x ≃ Ω X x :=
+{ to_fun := λ p, path.mk ⟨λ t, p (λ _, t), by continuity⟩
+    (gen_loop.boundary _ (λ _, 0) ⟨0, or.inl rfl⟩)
+    (gen_loop.boundary _ (λ _, 1) ⟨1, or.inr rfl⟩),
   inv_fun := λ p,
-  { to_fun := λ c, p c.head,
-    boundary := begin
-      rintro y ⟨i, iH|iH⟩; cases unique.eq_default i;
-      apply (congr_arg p iH).trans, exacts [p.source, p.target],
-    end },
-  left_inv := λ p, by { ext1, exact congr_arg p y.one_char.symm },
+  begin
+    refine ⟨⟨λ (c : I^fin 1), p (cube.head c), by continuity⟩, _⟩,
+    rintro y ⟨i, iH|iH⟩; cases unique.eq_default i;
+    apply (congr_arg p iH).trans, exacts [p.source, p.target],
+  end,
+  left_inv := λ p, by { ext, exact congr_arg p (cube.one_char y).symm },
   right_inv := λ p, by { ext, refl } }
 
-/--
-The first homotopy group at `x` is equivalent to the fundamental group,
-i.e. the loops based at `x` up to homotopy.
--/
-def pi1_equiv_fundamental_group : π 1 x ≃ fundamental_group X x :=
+/-- The first homotopy group at `x` is equivalent to the fundamental group, i.e. the loops based at
+  `x` up to homotopy. -/
+-- TODO: deduce from homotopy_group_equiv_fundamental_group?
+def pi1_equiv_fundamental_group : π_ 1 X x ≃ fundamental_group X x :=
```

Comment:
generalize to `unique` types

Reply:
After redefining `gen_loop_one_equiv_loop` the fact that it preserves `homotopy` should follow from earlier results.

Reply:
Oh even better: generalize `pi0_equiv_path_components` to empty types and use that + `homotopy_group_equiv_fundamental_group`

Reply:
Generalizations are done but I refrained from the refactoring because we should probably first combine [category_theory.functor.map_Aut](https://leanprover-community.github.io/mathlib_docs/category_theory/endomorphism.html#category_theory.functor.map_Aut) and [fundamental_groupoid.fundamental_groupoid_functor](https://leanprover-community.github.io/mathlib_docs/algebraic_topology/fundamental_groupoid/basic.html#fundamental_groupoid.fundamental_groupoid_functor) in another file, which is out of scope of this PR.

Reply:
@fpvandoorn I've done the refactoring in another branch anyway, and you can evaluate its (de)merits by [comparing both branches](https://github.com/leanprover-community/mathlib/compare/pi_n...homotopy_group_equiv_fundamental_group_of_unique_refactor?expand=1). Although defeq works nicely so that no `path.cast` is required, `@[simps]` times out at `gen_loop_equiv_of_unique` on the new branch. The whole refactoring saves 3 lines (not counting the lines added in path_connected that are not actually used).